src/lib/Libdb/db_postgres_job.c

```diff
@@ -405,60 +415,78 @@ load_job(const  PGresult *res, pbs_db_job_info_t *pj, int row)
-	SET_PARAM_INTEGER(conn, pjob->ji_ordering, 5);
-	SET_PARAM_INTEGER(conn, pjob->ji_priority, 6);
-	SET_PARAM_BIGINT(conn, pjob->ji_stime, 7);
-	SET_PARAM_BIGINT(conn, pjob->ji_endtBdry, 8);
-	SET_PARAM_STR(conn, pjob->ji_queue, 9);
-	SET_PARAM_STR(conn, pjob->ji_destin, 10);
-	SET_PARAM_INTEGER(conn, pjob->ji_un_type, 11);
-	SET_PARAM_BIGINT(conn, pjob->ji_momaddr, 12);
-	SET_PARAM_INTEGER(conn, pjob->ji_momport, 13);
-	SET_PARAM_INTEGER(conn, pjob->ji_exitstat, 14);
-	SET_PARAM_BIGINT(conn, pjob->ji_quetime, 15);
-	SET_PARAM_BIGINT(conn, pjob->ji_rteretry, 16);
-	SET_PARAM_INTEGER(conn, pjob->ji_fromsock, 17);
-	SET_PARAM_BIGINT(conn, pjob->ji_fromaddr, 18);
-	SET_PARAM_STR(conn, pjob->ji_4jid, 19);
-	SET_PARAM_STR(conn, pjob->ji_4ash, 20);
-	SET_PARAM_INTEGER(conn, pjob->ji_credtype, 21);
-	SET_PARAM_INTEGER(conn, pjob->ji_qrank, 22);
-
-	if (savetype == PBS_UPDATE_DB_QUICK) {
+
+	if (savetype & OBJ_SAVE_QS) {
+		SET_PARAM_INTEGER(conn, pjob->ji_state, 1);
+		SET_PARAM_INTEGER(conn, pjob->ji_substate, 2);
+		SET_PARAM_INTEGER(conn, pjob->ji_svrflags, 3);
+		SET_PARAM_INTEGER(conn, pjob->ji_numattr, 4);
+		SET_PARAM_INTEGER(conn, pjob->ji_ordering, 5);
+		SET_PARAM_INTEGER(conn, pjob->ji_priority, 6);
+		SET_PARAM_BIGINT(conn, pjob->ji_stime, 7);
+		SET_PARAM_BIGINT(conn, pjob->ji_endtBdry, 8);
+		SET_PARAM_STR(conn, pjob->ji_queue, 9);
+		SET_PARAM_STR(conn, pjob->ji_destin, 10);
+		SET_PARAM_INTEGER(conn, pjob->ji_un_type, 11);
+		SET_PARAM_BIGINT(conn, pjob->ji_momaddr, 12);
+		SET_PARAM_INTEGER(conn, pjob->ji_momport, 13);
+		SET_PARAM_INTEGER(conn, pjob->ji_exitstat, 14);
+		SET_PARAM_BIGINT(conn, pjob->ji_quetime, 15);
+		SET_PARAM_BIGINT(conn, pjob->ji_rteretry, 16);
+		SET_PARAM_INTEGER(conn, pjob->ji_fromsock, 17);
+		SET_PARAM_BIGINT(conn, pjob->ji_fromaddr, 18);
+		SET_PARAM_STR(conn, pjob->ji_4jid, 19);
+		SET_PARAM_STR(conn, pjob->ji_4ash, 20);
+		SET_PARAM_INTEGER(conn, pjob->ji_credtype, 21);
+		SET_PARAM_INTEGER(conn, pjob->ji_qrank, 22);
+
+		stmt = STMT_UPDATE_JOB_QUICK;
 		params = 23;
-	} else {
+	}
+
+	if ((pjob->db_attr_list.attr_count > 0) || (savetype & OBJ_SAVE_NEW)) {
 		int len = 0;
 		/* convert attributes to postgres raw array format */
-		if ((len = convert_db_attr_list_to_array(&raw_array, &pjob->attr_list)) <= 0)
+		if ((len = attrlist_2_dbarray(&raw_array, &pjob->db_attr_list)) <= 0)
 			return -1;
 
-		SET_PARAM_BIN(conn, raw_array, len, 23);
-		params = 24;
+		if (savetype & OBJ_SAVE_QS) {
+			SET_PARAM_BIN(conn, raw_array, len, 23);
+			params = 24;
+			stmt = STMT_UPDATE_JOB;
+		} else {
+			SET_PARAM_BIN(conn, raw_array, len, 1);
+			params = 2;
+			stmt = STMT_UPDATE_JOB_ATTRSONLY;
+		}
 	}
 
-	if (savetype == PBS_UPDATE_DB_FULL)
-		stmt = STMT_UPDATE_JOB;
-	else if (savetype == PBS_UPDATE_DB_QUICK)
-		stmt = STMT_UPDATE_JOB_QUICK;
-	else
+	if (savetype & OBJ_SAVE_NEW)
 		stmt = STMT_INSERT_JOB;
 
-	rc = pg_db_cmd(conn, stmt, params);
-
-	free(raw_array);
+	if (stmt != NULL) {
```

Comment:
Can the statement be NULL at this point? It does not seem like it can every be NULL. If it can be NULL, then maybe it would be better to move free(raw_array) out of this block.

Reply:
yes it could theretically be NULL, if it did not enter any of the above blocks - you are right, i move the free(raw_array) outisde